src/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -837,11 +837,6 @@ private SecurityStatusPal GenerateToken(byte[] input, int offset, int count, ref
             }
             finally
             {
-                if (_sslAuthenticationOptions.AlpnProtocolsHandle.IsAllocated)
-                {
-                    _sslAuthenticationOptions.AlpnProtocolsHandle.Free();
-                }
```

Comment:
Is the Close method that contains the remaining Free always going to be called?  e.g. is it called by both a Dispose/Close method and a finalizer?

Reply:
I believe its part of an Handle that is always called from my scanning of it.

Reply:
I'll be honest the pinning could be made tighter I suspect. Need to be careful though because I am not 100% the original supported renegotiation .... (Which there are no tests to catch). Because what happened if you freed the GCHandle, then called renegotiate and the callback was called? There was no catch if that was unallocated. However now it will last for the lifetime of the context.

Reply:
I just double checked that close is always called from SslStream.Dispose(disposing true) which comes from AuthenticatedStream which comes from Stream. I can't actually see a finalizer in that chain. So it might be missing finalisation. A better option might be to just hook into the SslContext handle as that already runs a finalizer... or to just add one to SslStream.

I can look at adding this to the SslContext as the callback is registered with that so freeing it just after the context is freed will ensure it's never run with an unpinned buffer.